Add examples for Notifications from async operations

h0lg opened the pull request:
Provides reproducable examples for the exceptions mentioned in #221 on the `NotificationPage` in the Gallery.

It also includes some **questions** regarding the recent changes towards the concept of **`CmdMsg`**, which I don't understand, seemingly replacing `Cmd<'msg>`, which I think I understand by now.

edgarfgp commented on samples/Gallery/Pages/NotificationsPage.fs:
```diff
@@ -31,16 +31,47 @@ module NotificationsPage =
+            return NotifyInfo "async operation completed"
+        }
+
+    let private notifyAsyncStatusUpdates dispatch =
```
In this case you will need to make sure the `Msg` is dispatched in the UI Thread. I think this will fix this.
```fsharp
let mutable cts: CancellationTokenSource = null

let private notifyAsyncStatusUpdates dispatch =
        if cts = null then
            cts <- new CancellationTokenSource()

            let notifications =
                async {
                    Dispatcher.UIThread.Post(fun _ -> dispatch(NotifyInfo "started"))
                    do! Async.Sleep 1000
                    Dispatcher.UIThread.Post(fun _ -> dispatch(NotifyInfo "5"))
                    do! Async.Sleep 1000
                    Dispatcher.UIThread.Post(fun _ -> dispatch(NotifyInfo "4"))
                    do! Async.Sleep 1000
                    Dispatcher.UIThread.Post(fun _ -> dispatch(NotifyInfo "3"))
                    do! Async.Sleep 1000
                    Dispatcher.UIThread.Post(fun _ -> dispatch(NotifyInfo "2"))
                    do! Async.Sleep 1000
                    Dispatcher.UIThread.Post(fun _ -> dispatch(NotifyInfo "1"))
                    do! Async.Sleep 1000
                    Dispatcher.UIThread.Post(fun _ -> dispatch(NotifyInfo "Completed"))
                }

            Async.Start(notifications, cts.Token)
        else
            cts.Cancel()
            cts.Dispose()
            cts <- null
```

h0lg replied:
Thanks! That did the trick for when I have access to a `dispatch` function by using `Cmd.ofEffect`, but not for my other example. See follow-up task.

edgarfgp commented:
Thanks for this PR.  I will go through it during the week and make some notes where we can add more docs or samples.

edgarfgp commented on samples/Gallery/Pages/NotificationsPage.fs:
```diff
@@ -73,13 +114,15 @@ module NotificationsPage =
 
             model, []
 
+        //TODO What is this about?
```
WindowNotificationManager can't be used immediately after creating it, so we need to wait for it to be attached to the visual tree. See https://github.com/AvaloniaUI/Avalonia/issues/5442

h0lg replied:
I've converted that into a comment.

edgarfgp commented on samples/Gallery/Pages/ClipboardPage.fs:
```diff
@@ -153,6 +155,7 @@ module ClipboardPage =
             return Cleared
         }
 
+    //TODO Why do this?
```
Strictly speaking this is only useful if you want to unit tests the commands, if not you can just use

```fsharp
Program.statefulWithCmd init update
```
Where the `init` and `update` functions will need to return `Cmd.none` or similars

h0lg replied:
I see, thanks for the explanation! That should probably be explained in detail in a prominent place, since this is all over the Gallery source code.

edgarfgp commented on samples/Gallery/Pages/NotificationsPage.fs:
```diff
@@ -50,6 +81,8 @@ module NotificationsPage =
         match msg with
         | ShowManagedNotification ->
 
+            //TODO this changes global state! i.e. after receiving this message, all notifications appear bottom right. is that intended?
```
Bear in mind that all of this is a port for the Avalonia Catalog samples, so I just wanted to test in the control works for us. have not really put too much though on making this `side-effect` free per se. Just adding some samples

h0lg replied:
I see. This was just something where the behaviour surprised me. Not so much after seeing the assignment `NotificationManager = FabApplication.Current.WindowNotificationManager` in the source - but it still raised the question of how I'd customize the position of just one message without doing the same for all following messages. Would I have to remember the original Position and reset it after `model.NotificationManager.Show`?

edgarfgp commented on samples/Gallery/Pages/NotificationsPage.fs:
```diff
@@ -93,6 +136,7 @@ module NotificationsPage =
 
     let view () =
         Component(program) {
+            //TODO What is this about?
```
We have added support for Mvu Components and now the `model` is not part of the `view` function instead you can access it thought `let! model = Mvu.State` like you will do in the `async` CE. We are planning to add some docs at some point once this is more stable for now this is in preview

edgarfgp replied:
I have created a separate repo `Fabulous.Avalonia.Samples` where I'm going create more small minimal samples ie  https://github.com/fabulous-dev/Fabulous.Avalonia.Samples/blob/main/MVU/CounterApp/App.fs

h0lg replied:
Thanks! I'll wait for those docs and will have a look at the samples.

h0lg commented:
@edgarfgp Thank you for taking the time to respond :) Your suggestion worked for one of the added scenarios examples. I have converted your feedback into hopefully helpful comments and updated my questions for you to have a look at when you get to it.

edgarfgp reviewed (approved):
Thanks for this PR. Improved the sample and took some notes to write some docs to clarify the `CmdMsg` and the new `Program` functions.
